core/src/main/java/com/radixdlt/genesis/olympia/converter/resource/OlympiaNonXrdConverter.java

```diff
@@ -0,0 +1,309 @@
+
+      var currentAllocations = ImmutableList.<GenesisResourceAllocation>builder();
+
+      for (var j = 0; j < balances.size(); j++) {
+        final var balance = balances.get(j);
+        final var olympiaAmount = balance.amount();
+        final var babylonAmount =
+            scaleResourceAmount(
+                olympiaAmount,
+                totalSupplyOnOlympia,
+                config.maxGenesisResourceUnscaledSupply().toRawBigIntRepresentation());
+
+        if (resourceTotalSuppliesOnBabylon[resourceIndex] == null) {
+          resourceTotalSuppliesOnBabylon[resourceIndex] = babylonAmount;
+        } else {
+          resourceTotalSuppliesOnBabylon[resourceIndex] =
+              resourceTotalSuppliesOnBabylon[resourceIndex].add(babylonAmount);
+        }
+
+        final var isLast =
+            i == olympiaNonXrdBalancesGroupedByResourceIdx.size() - 1 && j == balances.size() - 1;
+
+        final var account = accounts.get(balance.accountIndex());
+        final var accountAddress =
+            Address.virtualAccountAddress(account.publicKeyBytes().asBytes());
+
+        final var accountIndex =
+            accountsForCurrentResourceBalancesChunk.insertIfMissingAndGetIndex(accountAddress);
+
+        currentAllocations.add(
+            new GenesisResourceAllocation(UInt32.fromNonNegativeInt(accountIndex), babylonAmount));
+
+        numResourceBalancesInCurrentChunk += 1;
+        if (numResourceBalancesInCurrentChunk >= config.maxNonXrdResourceBalancesPerChunk()
+            || isLast) {
+          // We're over the per-chunk entries limit, or this is the last entry
+          // Build the chunk and add it to the list builder,
+          // reset the counter and the accounts index for the next chunk
+          resourceBalancesInCurrentChunkBuilder.add(
+              tuple(
+                  Address.globalFungible(padToResourceAddress(resource.addr().getBytes())),
+                  currentAllocations.build()));
+          currentAllocations = ImmutableList.builder();
+
+          resourceBalancesChunksBuilder.add(
+              new GenesisDataChunk.ResourceBalances(
+                  accountsForCurrentResourceBalancesChunk.build(),
+                  resourceBalancesInCurrentChunkBuilder.build()));
+          accountsForCurrentResourceBalancesChunk = new UniqueListBuilder<>();
+          resourceBalancesInCurrentChunkBuilder = ImmutableList.builder();
+          numResourceBalancesInCurrentChunk = 0;
+        }
+      }
+
+      // We've consumed all entries under the current resource key
+      // There might still be capacity for more entries (for different resources)
+      // So we're going to add an entry for the current resource
+      // and continue building the chunk
+      final var lastAllocationsForCurrentResource = currentAllocations.build();
+      if (!lastAllocationsForCurrentResource.isEmpty()) {
+        resourceBalancesInCurrentChunkBuilder.add(
```

Comment:
Hmm I might be missing something - but it looks like on the last loop, we never add the final chunk to `resourceBalancesChunksBuilder` before we build it?

EDIT: Ahh ignore me, that's handled by the `isLast` flag.